backend/src/auth/jwt.rs

```diff
@@ -0,0 +1,229 @@
+
+        jwt
+    }
+}
+
+impl JwtConfig {
+    fn load_auth(&self) -> Result<JwtAuth> {
+        let pem_encoded = std::fs::read(&self.secret_key)
+            .context("could not load secret key file")?;
+        let pem = pem::parse(pem_encoded)
+            .context("secret key file is not a valid PEM encoded key")?;
+
+        match self.signing_algorithm {
+            algo @ Algorithm::ES256 => JwtAuth::load_es(algo, &pem.contents),
+        }
+    }
+}
+
+
+
+struct JwtAuth {
+    signer: Box<dyn Signer>,
+    jwks: String,
+}
+
+impl JwtAuth {
+    /// Loads an elliptic curve key. `algo` has to be `ES256`!
+    fn load_es(algo: Algorithm, key: &[u8]) -> Result<JwtAuth> {
+        use elliptic_curve::pkcs8::DecodePrivateKey;
+
+        // Create a `ring` key pair that is used for signing.
+        let ring_algo = match algo {
+            Algorithm::ES256 => &ring::signature::ECDSA_P256_SHA256_FIXED_SIGNING,
+            // Algorithm::ES384 => &ring::signature::ECDSA_P384_SHA384_FIXED_SIGNING,
+        };
+        let ring_key = EcdsaKeyPair::from_pkcs8(ring_algo, key).map_err(|e| {
+            anyhow!("`jwt.secret_key` is not a valid ECDSA keypair in PKCS8 format: {}", e)
+        })?;
+
+        // Create the JWK(S) from the given key for the public route.
+        let jwk = match algo {
+            Algorithm::ES256 => <elliptic_curve::SecretKey<p256::NistP256>>::from_pkcs8_der(key)
+                .expect("failed to read ECDSA keypair, but it worked with `ring`?!")
+                .public_key()
+                .to_jwk(),
+        };
+
+
+        Ok(Self {
+            signer: Box::new(ring_key),
+            jwks: jwk_to_jwks(algo, jwk),
+        })
+    }
+}
+
+/// Serializes the given `jwk` from `elliptic_curve` into the expected JWKS structure.
+fn jwk_to_jwks(algo: Algorithm, jwk: impl Serialize) -> String {
+    #[derive(Serialize)]
+    struct Jwk<T: Serialize> {
+        #[serde(flatten)]
+        inner: T,
+
+        r#use: &'static str,
+        alg: &'static str,
+    }
+
+    #[derive(Serialize)]
+    struct Jwks<T: Serialize> {
+        keys: [Jwk<T>; 1],
+    }
+
+    let jwks = Jwks {
+        keys: [Jwk {
+            inner: jwk,
+            r#use: "sig",
+            alg: algo.to_str(),
+        }]
+    };
+    serde_json::to_string(&jwks).expect("failed to serialize JWKS")
+}
+
+/// A signature algorithm with corresponding key. Can sign a message.
+trait Signer: Sync + Send {
+    /// Signs the given message and writes the signature into `signature`.
+    fn sign(&self, rng: &dyn SecureRandom, message: &[u8], signature: &mut Vec<u8>);
+}
+
+impl Signer for EcdsaKeyPair {
+    fn sign(&self, rng: &dyn SecureRandom, message: &[u8], signature: &mut Vec<u8>) {
+        let sig = self.sign(rng, message).expect("failed to sign JWT");
+        signature.extend_from_slice(sig.as_ref())
+    }
+}
+
+
+// /// Now unused algorithm to convert ring key into valid PEM file. Only here to keep in
+// /// git history.
```

Comment:
Well, it's in the history now. But should it stay in the source as well? :thinking:

Reply:
Oh yeah wupps, I forgot to delete it :D